### PR TITLE
[#3905] Replace AsyncUsageAnalyzers with Microsoft.VisualStudio.Threading

### LINKED ISSUE #3905: Replace AsyncUsageAnalyzers with Microsoft.VisualStudio.Threading
https://www.nuget.org/packages/Microsoft.VisualStudio.Threading/ is the GA replacement of alpha https://www.nuget.org/packages/AsyncUsageAnalyzers/

[enhancement]

### PR DESCRIPTION
**Pull request by denscollo**
Fixes #3905

## Description

Replaces the [AsyncUsageAnalyzers](https://www.nuget.org/packages/AsyncUsageAnalyzers/) library with the latest version of its superseder [Microsoft.VisualStudio.Threading](https://www.nuget.org/packages/Microsoft.VisualStudio.Threading/).
All additional rules are explicitly suppressed in source, all pre-existing suppressions have been updated.
The following table describes all rule equivalences between both libraries:

|  Old | New | Description |
|---|---|---|
| [AvoidAsyncVoid](https://github.com/DotNetAnalyzers/AsyncUsageAnalyzers/blob/master/documentation/AvoidAsyncVoid.md) | [VSTHRD100](https://github.com/microsoft/vs-threading/blob/v16.8.55/doc/analyzers/VSTHRD100.md) | Avoid `async void` methods |
| [UseConfigureAwait](https://github.com/DotNetAnalyzers/AsyncUsageAnalyzers/blob/master/documentation/UseConfigureAwait.md) | [VSTHRD111](https://github.com/microsoft/vs-threading/blob/v16.8.55/doc/analyzers/VSTHRD100.md) | Use `.ConfigureAwait(bool)` |
| [AvoidAsyncSuffix](https://github.com/DotNetAnalyzers/AsyncUsageAnalyzers/blob/master/documentation/AvoidAsyncSuffix.md) & [UseAsyncSuffix](https://github.com/DotNetAnalyzers/AsyncUsageAnalyzers/blob/master/documentation/AvoidAsyncVoid.md) | [VSTHRD200](https://github.com/microsoft/vs-threading/blob/v16.8.55/doc/analyzers/VSTHRD200.md) | Use `Async` naming convention |
 
## Specific Changes

 - Replaces `AsyncUsageAnalyzers` with `Microsoft.VisualStudio.Threading.Analyzers` version `16.8.55` in `libraries/Directory.Build.props`
 - Suppresses all additional rules provided by the new library
 - Updates all pre-existing suppressions

## Testing
This is a screenshot of all the tests working with no errors for the changes introduced in this PR:
image

### CODE REVIEW

**Comment by mrivera-ms** [comment]
@gabog and @EricDahlvang to review.

**Comment by EricDahlvang** [comment]
Are the AsyncUsageAnalyzers rules in BotBuilder-DotNet.ruleset still used?

Also, it looks like AdaptiveExpressions.Tests is still referencing:

<PackageReference Include="AsyncUsageAnalyzers" Version="1.0.0-alpha003">

Can this be changed as well?

**Comment by ceciliaavila** [comment]
> Are the AsyncUsageAnalyzers rules in BotBuilder-DotNet.ruleset still used?
> 
> Also, it looks like AdaptiveExpressions.Tests is still referencing:
> 
> `<PackageReference Include="AsyncUsageAnalyzers" Version="1.0.0-alpha003">`
> 
> Can this be changed as well?

Hi @EricDahlvang,

Thanks for pointing this out to us, we reviewed these references and indeed, they were obsolete.
We removed them in our latest commit.

Thank you!

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Resources/ResourceExplorer.cs [hunk 1]
```diff
@@ -663,7 +665,9 @@ private void ResourceProvider_Changed(object sender, IEnumerable<Resource> resou
                     cancelReloadToken = new CancellationTokenSource();
 #pragma warning disable CA2008 // Do not create tasks without passing a TaskScheduler (this code looks problematic but excluding the rule for now, we need to analyze threading and re-entrance in more detail before trying to change it).
                     Task.Delay(1000, cancelReloadToken.Token)
+#pragma warning disable VSTHRD105 // Avoid method overloads that assume TaskScheduler.Current
```
Comment (gabog) [inline libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Resources/ResourceExplorer.cs:668]: Can we try to address this one based on the guidance in https://github.com/Microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD105.md
Leave the #pragma if you think the recommendation doesn't apply or is too risky

**Review by gabog — CHANGES_REQUESTED** [review]
Hi, sorry that it took so long for me to review this, we are at the beginning of R14 and I think it is a good time to fix VSTHRD103 (Call async methods when in an async method) rather than excluding it.  For example, replacing `streamWriter.FlushAsync();` by `await streamWriter.FlushAsync().ConfigureAwait(false);` is low risk and we can remove the #pragma.

The following rules are a bit more tricky and we should address them in a separate issue, for now, let's just add them as ignores in `BotBuilder-DotNet.ruleset` and remove the #pragmas from the code:

- [VSTHRD002](https://github.com/microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD002.md)
- [VSTHRD003](https://github.com/Microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD003.md)
- [VSTHRD011](https://github.com/microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD011.md)

**Review by gabog — APPROVED** [review]

**Pull request merged**